Update README.md (#14)
Anima-Tec/2021_Proyecto_Integrador_Equipo_3-Backend: README.md

Tambien se implemento el paquete [BackPack](https://backpackforlaravel.com/docs) de Laravel, que se usa como panel, donde el administrador controla todos los datos almacenados dentro de la Base de datos, pudiendo aprobarlos, rechazarlos, analizarlos y eliminarlos a gusto.

Esto es gracias a que BackPack nos da un panel pre definidos ya que este usa muchos paquetes y vistas que nos facilita su uso.

Para la creacion de las tablas y exportacion de datos dentro de la Base de Datos se utilizo _Migrations_ simplificando la exportacion de lo ya mencionado, estas crean un archivo que se puede modificar a gusto y se relaciona juntos al modelo utilizado MVC, sin neccesidad de hacer Querys de SQL.
 
## Deploy localmente

clonar el proyecto

```bash
  git clone https://github.com/Anima-Tec/2021_Proyecto_Integrador_Equipo_3-Backend.git
```

ir a la carpeta del proyecto
```bash
  cd 2021_Proyecto_Integrador_Equipo_3-Backend
```

instalar/actualizar composer

```bash
  composer install
  composer update
```

> usar/actualizar migrations

Reviewer: es necesario el `migrate:fresh`?

Reviewer: ```suggestion
correr las migraciones de nuestra base de datos
```

```bash
  php artisan migrate
  php artisan migrate:refresh
  ```

> Para correr el proyecto usar :

Reviewer: ```suggestion
Levantar el servidor :
```

```bash
  php artisan serve
```
---

> ## Variables de entorno

Reviewer: Pondria la secion de `Variables de entorno` antes que la de `Levantando el servicio`, ya que son necesaria para poder levantarlo.

Author: Me parece una buena observacion.

> Para poder correr el proyecto correctamente se necesita cambiar algunas variables de entorno, tales como :

Reviewer: ```suggestion
Para poder levantar el servicio correctamente se necesitan ciertas variables de entorno :
```

`DB_DATABASE`

`DB_USERNAME`

> `DB_PASSWORD`

Reviewer: Faltan mencionar varias variables de entorno, como la de los mails o las de mailtrap

Author: Estas seran remplazadas por el servicio que levantara facu de mails en un futuro, si esto no llegase a pasar, si se explicara el manejo de estas variables.

> Estas siendo para configurar donde se localizara la base de datos.

Reviewer: ```suggestion
Informacion sobre nuestra base de datos local.
```

---

## BackPack

Para la implementacion del Admin Panel se decidio utilizar Backpack, este siendo un conjunto de paquetes, el cual nos permite la moderacion del Backend.

Aplicando las funcionalidades requeridas para que un Administrador puede ver listadas las donaciones, peticiones, usuarios y en caso de aprobar una donacion o eliminarla este tenga las herramientas para hacerlo.

> 

Reviewer: No es necesario volver a nombrarlo, ya lo explicas arriba

Author: Me parecia bueno volver a traer el tema pero mas a fondo.

---

## Endpoints 

> A continuación se explicará todo sobre la creación de los Endpoints realizados y su formato dentro de la herramienta seleccionada [Postman](https://www.postman.com/).

Reviewer: ```suggestion
Encontraran una coleccion de [Postman](https://www.postman.com/) con todos los endpoints de nuestra aplicacion.
```

> ### Collection

Reviewer: ```suggestion
```

Author: Me parece que collections son neccesarias como titulo para explicar lo de Postaman

Archivo : proyecto_integrador_postman_collection

URL : https://www.postman.com/avionics-participant-66176292/workspace/proyecto-integrador-equipo-3

## Autenticación en los Endpoints

> Se utilizó el manejo de Tokens y de Rutas aseguradas para la Autenticación y Seguridad dentro del proyecto gracias a [Sanctum](https://laravel.com/docs/8.x/sanctum#how-it-works) que es un paquete mismo de Laravel.
> 
> Esto se decidio utilizar gracias a que se acoplaba al proyecto con sus neccesidades, por su facil forma de implementar dentro de Laravel.
> 

Reviewer: ```suggestion
La autentificacion de la aplicacion esta basada en tokens y la seguridad en rutas dependientes de la presencia de estos tokens. Para una correcta gestion de la autentificacion utilizamos [Sanctum](https://laravel.com/docs/8.x/sanctum#how-it-works).

Decidimos utilizar este metodo ya que se amolda a las necesidades del proyecto, ademas tiene cierta sencilles en su implementacion lo que agilizo el proceso de desarrollo.

```

---

> _Ejemplo de la protección vía Sanctum :_

Reviewer: ```suggestion
Las rutas que necesitan autentificacion estas agrupadas bajo el `middleware` de Sanctum dentro de nuestro archivo de rutas.
```

```php
  Route::group(['middleware' => ['auth:sanctum']], function () {

>     // Todas las rutas puestas aqui estarán protegidas y solo serán ejecutadas si la autenticación via token coincide.

Reviewer: ```suggestion
    // Todas las rutas puestas aqui estarán protegidas y solo podran ser accedidas si la autenticación via token es exitosa.
```

  });
```

> _Ejemplo de una ruta sin protección :_
> ```php
>   <Route::post('/DIRECCION_API', [CONTROLADOR::class, 'FUNCION_EJECUTAR']);
> ```

Reviewer: ```suggestion
Las rutas que esten por fuera del `middleware` no tendran autentificacion.
```

---

> ## Request

Reviewer: ```suggestion
## Paginacion en los endpoints
```

> Para la creacion de los endpoints dentro del proyecto, se utilizo la herramienta ya mencionada anteriormente, Postman.

Reviewer: ```suggestion
Para la creacion de los endpoints dentro del proyecto, se utilizo la herramienta ya mencionada anteriormente, Postman.
```
```suggestion

```

Para la facilitacion de obtencion de datos dentro delos endpoints, se decidio implementar _Pagination_ para que si reciben grandes cantidades de datos esto no sea un problema dentro de las Request, evitando un mal perfomance dividiendo el resultado obtenido en pequeñas listas.

> Aclarar que el limite de objetos recibidos a mostrar dentro de una Request es de 10, siendo lo suficientemente neccesario para el uso dentro del FrontEnd.

Reviewer: Hemos utilizado las facilidades que nos da Eloquent para poder paginar los endpoints de nuestra API que devuelven grandes cantidades de recursos. Decidimos que el limite de registros enviados en un endpoint paginado seria de 10, un limite que del lado del cliente es aceptable.

Author: Me parece bien.

> ### EJemplo de una Request con Pagination :

Reviewer: ```suggestion
### Ejemplo de una endpoint con paginacion :
```
